Iceberg split generation rate limit

marton-bod opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description
Hive, Hudi and DeltaLake connectors already use the same approach implemented in this PR, whereby they first asynchronously load (stream) splits into an `AsyncQueue` (which can be configured with throttling settings), and then when `getNextBatch()` is periodically called from the SplitSource, splits are borrowed from this queue. 

The primary aim is to control the rate at which splits are fetched for scheduling, thereby protecting the storage layer from getting overwhelmed by requests (e.g. namenodes in HDFS). This is achieved by a newly-introduced config flag (already present in Hive, Hudi and Delta): `iceberg.max-splits-per-second`.

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues
Solves the issue: https://github.com/trinodb/trino/issues/17974

The logic for filtering out splits should be _unchanged_: the pruning logic related to dynamic filter is in `getNextBatch()` since DF needs to be recomputed periodically. The rest of the pruning logic we can be applied to a split even before enqueueing it, therefore we can perform those checks within `acceptScanTask()` when we create the split stream.

Note: `FileAwareScanTask` is a new class introduced to keep metadata for FileScanTasks regarding their parent files. This is needed to replicate the current logic which stops the iteration if we hit enough records already for a LIMIT clause.

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(X) Release notes are required, with the following suggested text:

```markdown
# Iceberg connector
* Introduce new config property to rate limit split generation in Iceberg table scans: iceberg.max-splits-per-second.
```

marton-bod commented:
@alexjo2144 @findinpath Just a friendly bump, I would really appreciate it if you could review this PR when you get the chance. Thank you!

mosabua commented:
Given that this is modelled of implementation in other connectors this seems like a good approach. Could you potentially also look @electrum or @findepi

marton-bod commented:
Thanks a lot @mosabua for taking a first look. @electrum / @findepi please let me know if you have any thoughts, I do appreciate any input. Thanks a lot!

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:
```diff
@@ -73,6 +73,8 @@ public class IcebergConfig
     private Optional<String> materializedViewsStorageSchema = Optional.empty();
     private boolean sortedWritingEnabled = true;
+    private int maxSplitsPerSecond = Integer.MAX_VALUE;
+    private int maxOutstandingSplits = 1000;
```
document how this value was picked

marton-bod replied:
I picked this one to be consistent with Hive, Hudi and Delta, which use this same default value for `maxOutstandingSplits`

findepi replied:
```
// Default value picked after Hive, Hudi and Delta
```

marton-bod replied:
Comment added

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSessionProperties.java:
```diff
@@ -311,6 +313,16 @@ public IcebergSessionProperties(
                         icebergConfig.isSortedWritingEnabled(),
                         false))
+                .add(integerProperty(
+                        MAX_SPLITS_PER_SECOND,
+                        "Throttle the maximum number of splits that can be assigned to tasks per second",
+                        icebergConfig.getMaxSplitsPerSecond(),
+                        false))
+                .add(integerProperty(
+                        MAX_OUTSTANDING_SPLITS,
+                        "Target number of buffered splits for each table scan in a query, before the scheduler tries to pause itself",
+                        icebergConfig.getMaxOutstandingSplits(),
+                        false))
```
Not convinved we need this.
The PR quoted use-case (controlling HDFS namenode load) would ask for this not to be configurable by end user

marton-bod replied:
I've introduced this flag to be consistent with Hive, Delta and Hudi which all expose this config flag. Even though it's not the specific flag which addresses the PR-quoted use case, it was really easy to include this too in this PR, since it's needed for initializing the `AsyncQueue` anyway.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitManager.java:
```diff
@@ -44,19 +47,25 @@ public class IcebergSplitManager
     private final TrinoFileSystemFactory fileSystemFactory;
+    private final ExecutorService executorService;
     private final boolean asyncIcebergSplitProducer;
+    private final IcebergConfig config;
```
config is mutable, do not store it on a field

marton-bod replied:
Removed the field entirely, it's not needed

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitManager.java:
```diff
@@ -77,7 +86,7 @@ public ConnectorSplitSource getSplits(
 
         Table icebergTable = transactionManager.get(transaction, session.getIdentity()).getIcebergTable(session, table.getSchemaTableName());
-        Duration dynamicFilteringWaitTimeout = getDynamicFilteringWaitTimeout(session);
+        Duration dynamicFilteringWaitTimeout = config.getDynamicFilteringWaitTimeout();
```
why?

marton-bod replied:
This was left here as a mistake, restored the original state. Thanks!

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -115,17 +124,13 @@ public class IcebergSplitSource
     private final boolean recordScannedFiles;
     private final ImmutableSet.Builder<DataFileWithDeleteFiles> scannedFiles = ImmutableSet.builder();
     private long outputRowsLowerBound;
+    private boolean enoughSplits;
```
what is "enoughSplits"? the name hint at the  semantics to me

marton-bod replied:
Fair point. I have added a comment to describe its meaning. The main use of this flag is if there is a LIMIT clause: it indicates that no more splits are required for the scan, since we've already reached the required number of records to satisfy the LIMIT.

marton-bod replied:
I have removed `enoughSplits` actually. `FileAwareScanTask` will help enable the limit pushdown so this flag is no longer needed

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -174,134 +192,188 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
-                this.pushedDownDynamicFilterPredicate = TupleDomain.all();
-            }
+        if (taskGenerationException != null) {
+            return toCompletableFuture(immediateFailedFuture(taskGenerationException));
```
unnecessary double wrapping?

marton-bod replied:
Good catch. I initially took this pattern from the Delta implementation, but found a simpler way to avoid double wrapping.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -174,134 +192,188 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
-            Schema fileSchema = scanTask.spec().schema();
-            Map<Integer, Optional<String>> partitionKeys = getPartitionKeys(scanTask);
-
-            Set<IcebergColumnHandle> identityPartitionColumns = partitionKeys.keySet().stream()
```
the diff consist of new code being added and many lines moving around. It is hard for me to understand what are the important changes I should focus my review on.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -157,7 +156,6 @@ public IcebergSplitSource(
         this.fieldIdToType = primitiveFieldTypes(tableScan.schema());
         this.dynamicFilter = requireNonNull(dynamicFilter, "dynamicFilter is null");
         this.dynamicFilteringWaitTimeoutMillis = dynamicFilteringWaitTimeout.toMillis();
-        this.executor = executor;
```
should `Remove unused variable` commit be squashed?

marton-bod replied:
I'll squash all the commits into one

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -162,6 +172,16 @@ public IcebergSplitSource(
         this.fileModifiedTimeDomain = getFileModifiedTimePathDomain(tableHandle.getEnforcedPredicate());
+
+        // asynchronously collect the tasks and load them into the queue
+        this.queue = new ThrottledAsyncQueue<>(maxSplitsPerSecond, maxOutstandingSplits, executor);
```
How many threads there can be?

marton-bod replied:
It uses the iceberg executor service for processing, which is created using `Executors.newCachedThreadPool`. That has 0 threads for corePoolSize and Integer.MAX_VALUE for maxPoolSize.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergModule.java:
```diff
@@ -96,4 +102,11 @@ public void configure(Binder binder)
         tableProcedures.addBinding().toProvider(RemoveOrphanFilesTableProcedure.class).in(Scopes.SINGLETON);
     }
+
+    @Singleton
+    @Provides
+    public ExecutorService createIcebergExecutor()
+    {
+        return newCachedThreadPool(daemonThreadsNamed("iceberg-executor-%s"));
```
Who shuts this down on connector shutdown?

marton-bod replied:
Good catch. I think Delta is missing the shutdown step as well. I've changed the code to inject the iceberg executor service to the IcebergConnector so that it can be shut down when `IcebergConnector.shutDown()` is called

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -115,17 +125,13 @@ public class IcebergSplitSource
     private final Domain pathDomain;
     private final Domain fileModifiedTimeDomain;
     private final OptionalLong limit;
-
```
restore. this was separating immutable configuration fields from mutable fields responsible for tracking work progress

marton-bod commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetadataFileOperations.java:
```diff
@@ -181,32 +180,11 @@ public void testSelectWithLimit(int numberOfFiles)
-                        .addCopies(new FileOperation(SNAPSHOT, INPUT_FILE_NEW_STREAM), 1)
-                        .addCopies(new FileOperation(MANIFEST, INPUT_FILE_NEW_STREAM), numberOfFiles + min(icebergManifestPrefetching, numberOfFiles))
-                        .build());
+        assertFileSystemDataFileAccesses("SELECT * FROM test_select_with_limit LIMIT " + limit, min(limit, numberOfFiles));
```
@findepi the number of manifest accesses is no longer deterministic, because the splits are enqueued async. As soon as it's identified during a `getNextBatch` call that we have enough splits, the enqueuing will stop, but we can't make any guarantees about the number of manifest accesses.

On the other hand, the number of data file accesses are deterministic and no more data files will be read than necessary. Changed this test to reflect this.

findepi replied:
thinking about this

- we still want to test manifest accesses;  this is important metric
- there should probably be a kill-switch for this functionality

so maybe the test should just use that kill-switch?

marton-bod replied:
I've been thinking about this too, but I think adding a kill switch would complicate the code greatly. The iterative split generation approach is fundamentally different from the background async split loading approach, and a kill switch would require maintaining the code for both approaches in parallel and duplicating the logic in some places (and more painfully, porting fixes to both places). Because of this, I think it's impractical to introduce this flag.

I've made some enhancement to the test though, since even with this new approach we can make some guarantees about the number of manifest accesses. I've changed the test to compare the manifest access count against a lower bound and upper bound, and added some code comments to explain the reasoning behind the bounds. Regardless of the number of manifests in the scan, if there's a limit, the number of actually-accessed manifests should stay within a constant, narrow range.

findepi replied:
> The iterative split generation approach is fundamentally different from the background async split loading approach

kill switches are useful in case a bug is uncovered. if the new code  is fundamentally different from the old one, it sounds like it's not safe to proceed, is it?

marton-bod replied:
I phrased it wrong, I think. They're not fundamentally different, since pretty much all of the split generation and filtering logic has remained unchanged. It's different only in the sense that hiding it behind a feature flag is not trivial like putting some code in an if-else statement. But it's doable. For example, I could make two different classes `IcebergSplitSource` and `IcebergAsyncSplitSource` (with some common logic factored out into a parent base class) and toggle between the two implementations based on the flag - what do you think?

raunaqmorarka replied:
Can we make the new implementation produce the same behaviour as before if `max-outstanding-splits` was configurable and set to 1 ?

marton-bod commented:
@findepi Thanks so much for your time to review this. I've tried answering your questions as best as I could. Please let me know if you have any other questions/improvement ideas. Thanks!

marton-bod commented:
We have run TPC-DS 10000 (iceberg-parquet files in S3) with both our internal fork of 421, as well as 421 + this patch, and the results were identical (0.8% difference) cc @findepi @electrum

marton-bod commented:
@findepi @ebyhr @electrum  just a friendly check-in to see if you could please take another look at this PR? Would greatly appreciate any comments, opinions or guidance from all you. Thanks a lot!

vgankidi commented:
@findepi @ebyhr @electrum just a friendly check-in to see if you could please take another look at this PR?

mosabua commented:
Do we need commits squashed or something else done to progress this PR @findepi @electrum @martint ?

marton-bod commented:
@mosabua Thanks a lot for the ping! I have rebased the PR and squash the commits so it's ready to go again.
If the tagged maintainers are too busy, is there anyone else we can involve?

marton-bod commented:
@raunaqmorarka @findepi I have rebased to the latest master - can you please give this a review? It's been inactive for some time now so would be great to make some progress. Thank you!

vgankidi commented:
@findepi @ebyhr @electrum @raunaqmorarka just a friendly check-in to see if you could review the latest changes?

marton-bod commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -240,72 +208,124 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+        splits.add(icebergSplit);
+    }
+
+    private Optional<FileScanTask> getNextFileScanTask()
```
I broke this PR into two separate commits - the first one contains the method extraction: https://github.com/trinodb/trino/pull/18214/commits/52de00c0f88386134b9c760be13b7f6cfe99676d
Is that what you had in mind, or something different?

marton-bod replied:
Yeah, right now that `Prefactor` commit is the combination of extracting all 3 methods. Indeed not the easiest to follow - I will break it up as you suggested to make it easier to trace.

marton-bod commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergModule.java:
```diff
@@ -106,4 +112,11 @@ public void configure(Binder binder)
+
+    @Singleton
+    @Provides
+    public ExecutorService createIcebergExecutor()
```
Done

marton-bod commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -180,42 +203,90 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+                        fileTasksIterator = wholeFileTask.split(targetSplitSize).iterator();
+                    }
+                    List<FileAwareScanTask> subTasks = new ArrayList<>();
+                    boolean fileHasAnyDeletions = false;
```
You're right - simplified it

marton-bod commented:
@raunaqmorarka I've addressed your latest comments - please take another look to see if it's good to go now. Thanks!
The only test failure is unrelated (TestHiveTransactionalTable > testReadFullAcid)

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -99,6 +108,8 @@ public class IcebergSplitSource
     private static final ConnectorSplitBatch EMPTY_BATCH = new ConnectorSplitBatch(ImmutableList.of(), false);
     private static final ConnectorSplitBatch NO_MORE_SPLITS_BATCH = new ConnectorSplitBatch(ImmutableList.of(), true);
+    // It's a static value instead of a config, in order to protect users from setting this too high accidentally and causing coordinator memory issues
+    private static final int MAX_OUTSTANDING_SPLITS = 100;
```
100 outstanding splits sounds like a rather conservative value.

in fact, BufferingSplitSource by default uses buffer size of 100.
I don't know why we need BufferingSplitSource while we also have buffering inside the connector (and even once attempted to remove it https://github.com/trinodb/trino/pull/7394 but failed). Assuming we need both, Iceberg should have a bigger buffer size, and it should be configurable, jusjt like BufferingSplitSource is.

marton-bod replied:
Thanks @findepi. Other connector implementations, like [DeltaLakeConfig](https://github.com/trinodb/trino/blob/master/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java#L57) and [HudiConfig](https://github.com/trinodb/trino/blob/master/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiConfig.java#L49) set the default to 1000 for this value (and [Hive](https://github.com/trinodb/trino/blob/master/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveConfig.java#L77) even higher, to 3000). Initially, I also made this configurable and set the default to 1000 to be consistent with the other connectors. 

@raunaqmorarka had a concern that users might accidentally set this value too high and increase the memory consumption. Perhaps this is a relatively unlikely scenario if the other three connectors have not experienced this issue? Let me know if I should make this configurable again and set to 1000 default

raunaqmorarka commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -180,42 +205,85 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
     }
 
-    private void addScanTaskToSplits(TupleDomain<IcebergColumnHandle> dynamicFilterPredicate, List<ConnectorSplit> splits, FileScanTask scanTask)
+    private void addScanTaskToSplits(TupleDomain<IcebergColumnHandle> dynamicFilterPredicate, List<ConnectorSplit> splits, FileAwareScanTask fileAwareScanTask, AtomicBoolean reachedLimit)
```
Why don't we return a boolean here instead of updating an input parameter ?

marton-bod replied:
Good call

raunaqmorarka commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -180,42 +205,85 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+        executor);
+    }
+
+    private Stream<FileAwareScanTask> tasks()
```
tasks -> generateSplitScanTasks

raunaqmorarka commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -252,23 +320,22 @@ private void addScanTaskToSplits(TupleDomain<IcebergColumnHandle> dynamicFilterP
             scannedFiles.add(new DataFileWithDeleteFiles(scanTask.file(), fullyAppliedDeletes));
         }
-        if (!fileTasksIterator.hasNext()) {
+        if (fileAwareScanTask.isLastSplitOfParentFile()) {
```
Instead of adding isLastSplitOfParentFile/parentFileHasAnyDeletion and checking them from here, can we make `tasks()` method stop when "limit is reached" ? It seems to have all the required info to do this.

marton-bod replied:
The reason the limit is checked here, is that `tasks()` doesn't know which splits will actually be read and which will be filtered out by the dynamicFilter when the `getNextBatch()` is called. `tasks()` can only filter scanTasks  based on static information (in `filterScanTask()`), but dynamicFilter might still filter some of these scanTasks out later on

raunaqmorarka replied:
If limit is pushed into table scan, we don't want to apply dynamic filtering at all as that is just wasted work. In general we don't push predicate through limit, so a dynamic filter won't be present in the plan if a limit is present over the scan. Even if some future optimization were to push both dynamic filter and limit into the scan, we would want to ignore the dynamic filter in that case.

marton-bod commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -180,42 +205,85 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+                    List<FileAwareScanTask> subTasks = new ArrayList<>();
+                    while (fileTasksIterator.hasNext()) {
+                        FileScanTask scanTask = fileTasksIterator.next();
+                        filterScanTask(scanTask).ifPresent(task ->
```
You're absolutely right

marton-bod commented:
@raunaqmorarka Thanks again for reviewing this PR. I've addressed your latest comments. The tests are green, so it's ready for another pass. Couple of notes:
1. I have not turned `num-outstanding-splits` into a config yet, just increased its value to 1000. Let me know if you prefer to keep it as a static variable or a config. We can always introduce a config in a follow up PR too, if needed.
2. I have refactored the helper commits as per your suggestion into 3 separate commits (one for extracting a single method from `IcebergSplitSource`) - this should be much easier to follow now.

Please let me know if you have further comments. Thank you!

mosabua commented:
I assume you are still working on this @marton-bod and @raunaqmorarka  ... let me know if you need any help from others.

vgankidi commented:
I believe we need a re-review here.

mosabua commented:
@findepi @raunaqmorarka @electrum  could you help out here? also cc @brandylove

raunaqmorarka commented:
This needs a rebase to latest master first

marton-bod commented:
@raunaqmorarka I have rebased to the latest master. Please take a look again, it should be good to go now. 
Thanks a lot!

marton-bod commented:
@raunaqmorarka I've rebased this patch once again. Please take a look when you have a chance - thank you.

marton-bod commented:
Can anyone please review this? I've been having to rebase this patch again and again for a very long time now. Thank you.

bitsondatadev commented:
@raunaqmorarka @findepi @findinpath

findepi commented:
@electrum

vgankidi commented:
@raunaqmorarka @findepi @findinpath @electrum Would really appreciate reviews here, its been a long pending one. Hope to close this out soon.

vgankidi commented:
@raunaqmorarka @findepi @findinpath @electrum @mosabua   Gentle ping again on this.

bitsondatadev commented:
@vgankidi  I'm just gonna ping and ping until someone can get to this lol. Sorry this has taken so long.

dain commented:
This feature is pretty anti Trino.  Trino is designed to complete queries as fast as possible.  Storage systems should be protecting themselves by pushing back and throttling (like all of the cloud ones do).  Also split generation is not representitive of actually performing any IO, so even if we wanted to add a throttle, this is the wrong place.

dain reviewed (changes requested):
I don't think we should be doing this, and it should be discussed (slack, contributors meeting, etc.) before merging.

dain commented:
@electrum pointed out this is in the outher Hive style connectors already.  I have no idea why this was added because it is a poor stand in for real concurrency control... anyway, I suggest we do not document or recommend this feature.

marton-bod commented:
@dain 
- The same approach is already followed by both Hive, Delta Lake and Hudi. As for the "anti-trinoness" of this approach, I would just remind you  of this Delta commit: https://github.com/trinodb/trino/commit/9956543e, which introduces the Delta connector, which was co-authored by _you_ as well as couple dozen Starburst engineers, signing off on this very same approach. 
- The point of this change is to protect storage systems which do not have throttling capabilities, like HDFS. I clearly pointed this out in the [issue description](https://github.com/trinodb/trino/issues/17974) as well. We have plenty of Iceberg datasets in HDFS. For storage systems with throttling, like S3, I agree this shouldn't be used, but no one suggested it otherwise.

dain commented:
I suggest we fix this up and merge it, so there is parity across the implementations.  Then I suggest we add a real throttling wrapper to the new native file systems to actualy deal with real errors (we can start with a simpel per machine uncoordinated limt).  Once that is in we should drop this from all of the connectors.

mosabua commented:
Awesome @dain .. do you want me to file an issue so we can track this and move it along?

@marton-bod can you rebase and ping us when its ready for review again? Then we can figure out how @findepi @raunaqmorarka @ebyhr and @dain collaborate efficiently to proceed this to merge.

marton-bod commented:
@mosabua @dain Thanks all! I've rebased this to the latest master, so should be ready again for reviews

mosabua commented:
Please rebase @marton-bod 

Either @findepi or @dain can then proceed with final review and merge

marton-bod commented:
@mosabua @dain Rebase done. Thanks!

marton-bod commented:
@mosabua Shall we just close this PR instead?

mosabua commented:
@marton-bod @dain .. we kinda agreed to get this merged to get parity for all the object storage connectors on this feature. I still think we should do that .. and also work on a proper solution, maybe on the file system level @electrum 

Thoughts?
